### PR TITLE
Improved Command Line with CLI11

### PR DESCRIPTION
**Pull request by st4rl3ss**
Implemented the CLI11 as new command line parser as requested in #150 .

### CODE REVIEW

**Comment by pramodk** [comment]
I haven't looked at details but could you also update readme?

(@iomaganaris also take a look when you time permits.)

File: coreneuron/nrniv/main1.cpp [hunk 1]
```diff
@@ -368,15 +372,27 @@ const char* nrn_version(int) {
 
 using namespace coreneuron;
 
-extern "C" void mk_mech_init(int argc, char** argv) {
-#if NRNMPI
-    nrnmpi_init(1, &argc, &argv);
-#endif
+
+extern "C" int mk_mech_init(int argc, char** argv) {
```
Comment (iomaganaris) [inline coreneuron/nrniv/main1.cpp]: Small detail.. I did not understand why `mk_mech_init()` should be `int` since it doesn't return any value

File: coreneuron/nrniv/cn_parameters.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "coreneuron/nrniv/cn_parameters.h"
+
+cn_parameters::cn_parameters(){
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.cpp]: Nitpick.. Regarding the `cn_` naming, I think I would prefer something like `corenrn`. What do you think? You can see [here](https://github.com/BlueBrain/hpc-coding-conventions/blob/master/cpp/NamingConventions.md#use-descriptive-variable-names) also some suggestions about the naming of variables, classes, etc

File: coreneuron/nrniv/cn_parameters.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    app.get_formatter()->column_width(50);
+    app.set_help_all_flag("-H, --help-all", "Print this help including subcommands and exit.");
+
+    app.set_config("--config", "config.ini", "Read parameters from ini file", false)
+        ->check(CLI::ExistingFile);
+    app.add_flag("--mpi", this->mpi_en, "Enable MPI. In order to initialize MPI environment this argument must be specified." );
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.cpp]: Same as above, `mpi_en` could be `mpi_enable`

File: coreneuron/nrniv/cn_parameters.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        ->check(CLI::ExistingPath);
+    sub_input -> add_option("-f, --filesdat", this->filesdat, "Name for the distribution file.", true)
+        ->check(CLI::ExistingFile);
+    sub_input -> add_option("-p, --pattern", this->patternstim, "Apply patternstim using the specified spike file.")
+        ->check(CLI::ExistingFile);
+    sub_input -> add_option("-s, --seed", this->seed, "Initialization seed for random number generator.")
+        ->check(CLI::Range(0, 100000000));
+    sub_input -> add_option("-v, --voltage", this->voltage, "Initial voltage used for nrn_finitialize(1, v_init). If 1000, then nrn_finitialize(0,...).")
+        ->check(CLI::Range(-1e9, 1e9));
+    sub_input -> add_option("--read-config", this->rconfigpath, "Read configuration file filename.")
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.cpp]: A small naming detail suggestion again. If `rconfigpath` is a file maybe it should be named as `rconfigfilepath` to avoid any confusion between the path where the file is and the file. Same with `restorepath`

File: coreneuron/nrniv/cn_parameters.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    string_t patternstim;          /// Apply patternstim using the specified spike file.
+    string_t datpath=".";          /// directory path where .dat files
+    string_t outpath=".";          /// directory where spikes will be written
+    string_t filesdat="files.dat"; /// name of file containing list of gids dat files read in
+    string_t rconfigpath;
+    string_t restorepath;
+    string_t reportpath;
+    string_t checkpointpath;
+
+    CLI::App app{"CoreNeuron - Your friendly neuron simulator."}; ///CLI app that performs CLI parsing
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.h]: Nice title :laughing:

File: coreneuron/nrniv/cn_parameters.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+*/
+
+#ifndef CN_PARAMETERS_H
+#define CN_PARAMETERS_H
+
+#include <string>
+#include <fstream>
+#include <iostream>
+#include <iomanip>
+#include "coreneuron/utils/CLI11/CLI.hpp"
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.h]: Maybe it would be better to add `CLI11` as submodule, similarly to [`NMODL`](https://github.com/BlueBrain/nmodl/blob/master/.gitmodules) and include the header file from the `CLI11` repository rather than the `CLI.hpp` header file in this project. @pramodk what do you think about that?

File: coreneuron/nrniv/cn_parameters.h [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+ * the command line parameters for subcategories NEED to be come before the relative
+ * parameter. e.g. --mpi --gpu gpu --nwarp
+ * Also single dash long options are not supported anymore (-mpi -> --mpi).
+ */
+
+typedef std::string string_t;
+
+struct cn_parameters {
+
+    int spikebuf=100000;           /// internal buffer used on every rank for spikes
```
Comment (iomaganaris) [inline coreneuron/nrniv/cn_parameters.h]: I think it would be nice all the variables to have descriptions in the same place

File: coreneuron/nrniv/main1.cpp [hunk 8]
```diff
@@ -261,43 +265,43 @@ void nrn_init_and_load_data(int argc,
...
+    if (!cn_par.patternstim.empty()) {
+        nrn_mkPatternStim(cn_par.patternstim.c_str());
     }
 
     /// Setting the timeout
     nrn_set_timeout(200.);
 
     // show all configuration parameters for current run
-    nrnopt_show();
+    //nrnopt_show();
```
Comment (iomaganaris) [inline coreneuron/nrniv/main1.cpp]: The commented line can be deleted if it is not used any more

File: coreneuron/nrniv/main1.cpp (hunk 1 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: For whatever reason, if I let it `void`, the compilation would just fail. I'll check it out again.

File: coreneuron/nrniv/cn_parameters.cpp (hunk 2 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.cpp]: Yeah, maybe `cn` is too concise. There is no official abbreviation for Coreneuron or I would have used that. `corenrn` seems a good compromise.

File: coreneuron/nrniv/cn_parameters.cpp (hunk 3 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.cpp]: In the HW world nobody would discuss the meaning of `en` but I guess a bit more clarity wouldn't hurt. I'll change as you suggested.

Comment (iomaganaris) [reply coreneuron/nrniv/cn_parameters.cpp]: Haha true :+1:

File: coreneuron/nrniv/cn_parameters.cpp (hunk 4 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.cpp]: Fair enough!

File: coreneuron/nrniv/cn_parameters.h (hunk 5 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.h]: 😄 Now the question is: should I remove it? Or leave it?

File: coreneuron/nrniv/cn_parameters.h (hunk 6 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.h]: That seems like a great idea! We can do this in the afternoon.

File: coreneuron/nrniv/cn_parameters.h (hunk 7 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.h]: Not all the variables had a description and for most of them I had no idea what they did. I can certainly ask and complete the descriptions.

File: coreneuron/nrniv/main1.cpp (hunk 8 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: My bad!

File: coreneuron/nrniv/cn_parameters.h (hunk 5 above)
Comment (iomaganaris) [reply coreneuron/nrniv/cn_parameters.h]: Haha in the [README.md](https://github.com/BlueBrain/CoreNeuron#coreneuron) there is a title `Optimised simulator engine for NEURON`. Maybe this one should be used in the cli too?

Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.h]: Of course we should use that. But mine was funnier!

File: coreneuron/nrniv/cn_parameters.h (hunk 7 above)
Comment (iomaganaris) [reply coreneuron/nrniv/cn_parameters.h]: I see.. You can add whatever you can find :+1:

File: coreneuron/nrniv/cn_parameters.h (hunk 5 above)
Comment (iomaganaris) [reply coreneuron/nrniv/cn_parameters.h]: There is no double for that :+1:

File: coreneuron/nrniv/main1.cpp [hunk 9]
```diff
@@ -453,32 +470,33 @@ extern "C" int run_solve_core(int argc, char** argv) {
         }
         // Set the buffer size if is not the default value. Otherwise use report.conf on
         // register_report
-        if (!nrnopt_is_default_value("--report-buffer-size")) {
+        if (cn_par.report_buff_size!=4) {
```
Comment (iomaganaris) [inline coreneuron/nrniv/main1.cpp]: I think that testing if the `report_buff_size` is equal to the number `4` is not a very good idea, since the default value may change in the future. If there is not any better way to check if an option is given in `CLI11` you could use an extra member with the default `buff_size` and check the `cn_par.report_buff_size` with it at least

Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: Probably not the smartest idea indeed. Thanks for your advice, I'll find a better solution to this.

File: coreneuron/nrniv/main1.cpp [hunk 10]
```diff
@@ -368,15 +372,27 @@ const char* nrn_version(int) {
...
+extern "C" int mk_mech_init(int argc, char** argv) {
+    #if NRNMPI
+        nrnmpi_init(1, &argc, &argv);
+    #endif
     // read command line parameters and parameter config files
-    nrnopt_parse(argc, (const char**)argv);
+
+    cn_par.parse(argc, argv);
+
+    if (cn_par.print_arg) {
```
Comment (iomaganaris) [inline coreneuron/nrniv/main1.cpp]: I was wondering if the parameters should be always printed, since this is what was happening before. I also got some issues with printing the parameters two times and some of them being printed wrong when running for example `tests/integration/ring/integration_test.sh` with `--show`. Do you have the same problem?

Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: I kept the same settings as the previous implementation, so they get printed every time as a confirmation of what is happening. I'll check what happen in the case you described.

File: tests/integration/CMakeLists.txt [hunk 11]
```diff
@@ -25,32 +25,34 @@
...
-set(PERMUTE2_ARGS "--cell-permute 2 --nwarp 16")
+set(RING_COMMON_ARGS "${COMMON_ARGS} input --datpath ${RING_DATASET_DIR}")
+set(RING_SPIKEBUFFER_ARGS "${COMMON_ARGS} -b 1 input --datpath ${RING_DATASET_DIR}")
+set(RING_DUPLICATE_ARGS "${COMMON_ARGS} -z 3 input --datpath ${RING_DATASET_DIR}")
+set(RING_GAP_COMMON_ARGS "${COMMON_ARGS} input --datpath ${CMAKE_CURRENT_SOURCE_DIR}/ring_gap")
+set(PERMUTE1_ARGS "gpu --cell-permute 1 --nwarp 8")
+set(PERMUTE2_ARGS "gpu --cell-permute 2 --nwarp 16")
 if(ENABLE_OPENACC)
-  set(GPU_ARGS "--nwarp 8 --gpu")
+  set(GPU_ARGS "gpu --nwarp 8 --gpu")
```
Comment (iomaganaris) [inline tests/integration/CMakeLists.txt]: There seems to be some kind of issue with `--gpu` coming probably from here that fails the [jenkins_plan](https://bbpcode.epfl.ch/ci/blue/organizations/jenkins/oss.coreneuron.pipeline/detail/oss.coreneuron.pipeline/379/pipeline). We can check it together also at some point

File: coreneuron/nrniv/cn_parameters.h (hunk 5 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/cn_parameters.h]: I'm sad now.

File: coreneuron/nrniv/corenrn_parameters.h [hunk 12]
```diff
@@ -33,7 +33,7 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include <fstream>
 #include <iostream>
 #include <iomanip>
-#include "coreneuron/utils/CLI11/CLI.hpp"
+#include "external/CLI11/include/CLI/CLI.hpp"
```
Comment (iomaganaris) [inline coreneuron/nrniv/corenrn_parameters.h]: Does the following work too?
```suggestion
#include "CLI/CLI.hpp"
```

File: tests/unit/cmdline_interface/test_cmdline_interface.cpp [hunk 13]
```diff
@@ -32,7 +32,7 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include <boost/test/unit_test.hpp>
 #include "nrniv/nrnoptarg.h"
 #include <float.h>
-#include <utils/CLI11/CLI.hpp>
+#include "external/CLI11/include/CLI/CLI.hpp"
```
Comment (iomaganaris) [inline tests/unit/cmdline_interface/test_cmdline_interface.cpp]: Same as above:
```suggestion
#include "CLI/CLI.hpp"
```

Comment (iomaganaris) [reply tests/unit/cmdline_interface/test_cmdline_interface.cpp]: Perfect :ok_hand:

File: coreneuron/nrniv/main1.cpp [hunk 14]
```diff
@@ -378,11 +378,13 @@ extern "C" void mk_mech_init(int argc, char** argv) {
     #endif
     // read command line parameters and parameter config files
 
-    corenrn_param.parse(argc, argv);
+    try {corenrn_param.parse(argc, argv);} catch (...) {exit(1);}
```
Comment (iomaganaris) [inline coreneuron/nrniv/main1.cpp]: The addition of exceptions is pretty nice! :+1: 
I would only suggest to use [`nrn_abort()`](https://github.com/BlueBrain/CoreNeuron/blob/f7d78fb83197d946f506ecaedbfafb8a64d362ac/coreneuron/nrnmpi/nrnmpi.cpp#L192) instead of `exit()` to make sure that all the MPI processes exit and also to be a bit cleaner since `nrnmpi_init()` is already called. 
It would also be nice to reformat this line like this:
```suggestion
try {
   corenrn_param.parse(argc, argv);
}
catch (...) {
   nrn_abort(1);
}
```

Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: Nice "Catch", Ioannis! See? I made a nerdy joke! hahaha, that was funny! Anyway, if `nrn_abort(1)` is functionally equivalent to `exit(1)` it's definitely the better option.

Comment (iomaganaris) [reply coreneuron/nrniv/main1.cpp]: Hahaha noooo :laughing: 
It should be the same functionality, yes. In the implementation of `MPI_Abort()` there is probably going to be called `exit()` at some point but I think that it is better to use the `nrn_<mpi_function>()` functions that call the related `MPI functions` for the MPI operations. You can read more about  `MPI_Abort()` [here](https://www.mpich.org/static/docs/v3.1/www3/MPI_Abort.html) and [here](https://stackoverflow.com/questions/54844522/mpi-abort-vs-exit).

Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: Should be `nrn_abort()` then. I'll check it out. Thanks!

File: README.md [hunk 15]
```diff
@@ -140,20 +140,22 @@ mpiexec -np 2 build/apps/coreneuron_exec -e 10 --mpi input -d /path/to/model/bui
...
+Some details on the new interface:
 
-**The new command line interface is based on CLI11. All the previous options are still supported but they are organized in subcommands. You can find more details by running `coreneuron_exec --help-all`.**
+The new command line interface is based on CLI11. All the previous options are still supported but they are organized in subcommands. You can find more details by running `coreneuron_exec --help-all`.
 
-**Also multiple character options with single dash (e.g. `-gpu`) are not supported anymore. All multiple characters options require a double dash now (e.g. `--gpu`).**
+Also multiple character options with single dash (e.g. `-gpu`) are not supported anymore. All multiple characters options require a double dash now (e.g. `--gpu`).
 
-**Since commands are now organized in subcommands, to access a certain parameter it might be necessary to include in the command line a certain keyworld first.**
+Since commands are now organized in subcommands, to access a certain parameter it might be necessary to include in the command line a certain keyworld first.
```
Comment (iomaganaris) [inline README.md]: I feel that it should be more helpful to state clearer that the `options` should always be before the subcommands

File: README.md [hunk 16]
```diff
@@ -140,20 +140,22 @@ mpiexec -np 2 build/apps/coreneuron_exec -e 10 --mpi input -d /path/to/model/bui
...
 
-**For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:**
-**`coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron`** 
+For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:
+`coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron`
 
-**Multiple options in a certain subcommand section can be specifying the keyword only once: `coreneuron_exec --mpi input -d /path/to/model -f /path/to/filesdat/file.dat`**
+Multiple options in a certain subcommand section can be specifying the keyword only once: `coreneuron_exec --mpi input -d /path/to/model -f /path/to/filesdat/file.dat`
```
Comment (iomaganaris) [inline README.md]: Maybe this could be rephrased a bit?
```suggestion
To define **more than one** option in a certain subcommand it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand (for example `-d /path/to/model -f /path/to/filesdat/file.dat`). The full command would be:
`coreneuron_exec --mpi input -d /path/to/model -f /path/to/filesdat/file.dat`
```
Or something similar? Feel free to modify my suggestion if you want

**Comment by st4rl3ss** [comment]
please retest

**Comment by st4rl3ss** [comment]
please retest

File: README.md [hunk 17]
```diff
@@ -131,63 +131,103 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+
+:warning: :warning: :warning: **In a recent update the command line interface was updated, so please update your scripts accordingly!**
+
+Some details on the new interface:
+
+The new command line interface is based on CLI11. All the previous options are still supported but they are organized differently. You can find more details by running `coreneuron_exec --help-all`.
+
+Multiple character options with single dash (e.g. `-gpu`) are not supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters option still support a single dash (e.g. `-g`).
+
+The structure of a command is the following: `./apps/coreneuron_exec [OPTIONS] [SUBCOMMAND]`. The `OPTIONS` should **always** be written before any of the `SUBCOMMAND`. To know which of the parameter is an `OPTION` or a `SUBCOMMAND` run `coreneuron_exec --help-all`.
```
Comment (ohm314) [inline README.md]: Do you mean by `OPTIONS` here general options that don't belong to any subcommand? This should be clarified. The current text seems to contradict below paragraph.

File: README.md [hunk 18]
```diff
@@ -131,63 +131,103 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+
+To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyworld first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:
```
Comment (ohm314) [inline README.md]:
```suggestion
To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyword first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword first:
```

Comment (ohm314) [reply README.md]: Also, the paragraph should be updated for clarity. It's not really clear, in this form.

File: README.md [hunk 19]
```diff
@@ -131,63 +131,103 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+`coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron`
+
+To define **more than one** option in a certain subcommand it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand (for example `-d /path/to/model -f /path/to/filesdat/file.dat`). The full command would be:
```
Comment (ohm314) [inline README.md]: You seem to use _option_ and _parameter_ interchangeably. This is confusing. Also this paragraph and the previous can be written shorter and as one.

File: coreneuron/nrniv/corenrn_parameters.cpp [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "coreneuron/nrniv/corenrn_parameters.h"
+
+corenrn_parameters::corenrn_parameters(){
+
```
Comment (ohm314) [inline coreneuron/nrniv/corenrn_parameters.cpp]: Just wondering, is there an easy way to check if the user used legacy options such as `-mpi` and fail with a meaningful message?

File: coreneuron/nrniv/corenrn_parameters.cpp [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+        ->check(CLI::ExistingPath);
+    sub_output -> add_option("--checkpoint", this->checkpointpath, "Enable checkpoint and specify directory to store related files.")
+        ->check(CLI::ExistingDirectory);
+};
+
+void corenrn_parameters::parse (int argc, char** argv) {
+
+    try {
+        app.parse(argc, argv);
+    } catch (const CLI::ParseError &e) {
```
Comment (ohm314) [inline coreneuron/nrniv/corenrn_parameters.cpp]: Would calling with `-mpi` cause a ParseError? If so, maybe the catch should print out a message to tell the user what exactly is the issue (that we moved from single-dash to double-dash long-parameter syntax).

File: coreneuron/nrniv/main1.cpp [hunk 22]
```diff
@@ -387,42 +409,43 @@ extern "C" int run_solve_core(int argc, char** argv) {
...
+                                                   corenrn_param.outpath.c_str());
             reports_needs_finalize = configs.size();
         }
     }
 
-    // initializationa and loading functions moved to separate
+    // initialization and loading functions moved to separate
     Instrumentor::phase_begin("load-model");
-    nrn_init_and_load_data(argc, argv, !configs.empty());
+    nrn_init_and_load_data(argc, argv, configs.size() > 0);
```
Comment (ohm314) [inline coreneuron/nrniv/main1.cpp]: why did you change this?

File: README.md (hunk 17 above)
Comment (st4rl3ss) [reply README.md]: That's what I mean! It's pretty unclear so I'll clarify this...

File: README.md (hunk 19 above)
Comment (st4rl3ss) [reply README.md]: I tried to use the word `parameter` in a generic way, while using `option` for general options that do not belong to any subcommands. In a few places it was inconsistent so I fixed it.

File: README.md (hunk 18 above)
Comment (st4rl3ss) [reply README.md]: By including the definition of `options` as parameters that do not belong to any of the `subcommands` it should be a bit clearer. But I understand it can be a bit foggy. What parts in particular would you improve? What things are not clear for you?

File: coreneuron/nrniv/corenrn_parameters.cpp (hunk 20 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/corenrn_parameters.cpp]: As far as I've read in the CLI11 documentation, there is no way to implement something like that easily. But it would be a welcome addition indeed.

Maybe I can add a manual check  on some particular common options that wouldn't work in future versions...

File: coreneuron/nrniv/corenrn_parameters.cpp (hunk 21 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/corenrn_parameters.cpp]: See above. It would be a great feature. I'll try to add it.

File: coreneuron/nrniv/main1.cpp (hunk 22 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: I think it was a suggestion from @pramodk at some point during development. Is there anything wrong about it? Do you think it's an improvement?

File: coreneuron/nrniv/corenrn_parameters.cpp (hunk 21 above)
Comment (ohm314) [reply coreneuron/nrniv/corenrn_parameters.cpp]: see if `CLI::ParseError` can provide the necessary information.

File: coreneuron/nrniv/main1.cpp (hunk 22 above)
Comment (ohm314) [reply coreneuron/nrniv/main1.cpp]: generally we prefer to use `empty()` over `size() == 0` since the former is more literal and shorter.

File: README.md [hunk 23]
```diff
@@ -131,63 +131,100 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
 
+### Command Line Interface
+
+:warning: :warning: :warning: **In a recent update the command line interface was updated, so please update your scripts accordingly!**
+
+Some details on the new interface:
+
+The new command line interface is based on CLI11. All the previous options are still supported but they are organized differently. You can find more details by running `coreneuron_exec --help-all` or using the table reported below.
+
+Multiple character options with single dash (e.g. `-gpu`) are **not** supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters option still support a single dash (e.g. `-g`).
```
Comment (ohm314) [inline README.md]:
```suggestion
Multiple character options with single dash (e.g. `-gpu`) are **not** supported anymore. All multiple characters options now require a double dash (e.g. `--gpu`), but single characters options still support a single dash (e.g. `-g`).
```

File: README.md [hunk 24]
```diff
@@ -131,63 +131,100 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+
+The structure of a command is the following: `./apps/coreneuron_exec [OPTIONS] [SUBCOMMAND]`. More specifically, an `OPTION` is a parameter that doesn't belong to **any** `SUBCOMMANDS`. The `OPTIONS` should **always** be written before any of the `SUBCOMMAND`. To know which of the parameter is an `OPTION` or a `SUBCOMMAND` run `coreneuron_exec --help-all` or refer to the table below.
```
Comment (ohm314) [inline README.md]:
```suggestion
The structure of a command is the following: `./apps/coreneuron_exec [GENERIC OPTIONS] [SUBCOMMAND [OPTIONS]]`. More specifically, a `GENERIC OPTION` is a parameter that doesn't belong to **any** `SUBCOMMANDS`. `GENERIC OPTIONS` should **always** be written before any `SUBCOMMAND`. To know which of the various subcommands and their respective options run `coreneuron_exec --help-all` or refer to the table below.
```

File: README.md [hunk 25]
```diff
@@ -131,63 +131,100 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+
+To access a parameter that falls under any of the subcommands category it is necessary to include in the command line a certain keyword first. For example to use the `-d` option, one must enter in the `input` subcommands first by using the `input` keyword:
```
Comment (ohm314) [inline README.md]:
```suggestion
To access a subcommand-specific parameter it must always follow that subcommand. For example to use the `-d` option, one must first write the `input` subcommand.
```

File: README.md [hunk 26]
```diff
@@ -131,63 +131,100 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
+therefore the old command `coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron` in the new implementation. If you want to  define **more than one** parameter in a certain `SUBCOMMAND` it is needed to state the **subcommand keyword** (for example `input`) **directly followed** by the **options** of this subcommand. The `SUBCOMMAND` keyword needs to be stated only once for every set of parameters.
```
Comment (ohm314) [inline README.md]:
```suggestion
Hence, the old command `coreneuron_exec -d /path/to/model/built/by/neuron` becomes `coreneuron_exec input -d /path/to/model/built/by/neuron` in the new implementation. All subcommand-specific options must be grouped with their respective subcommand.
```

File: coreneuron/nrniv/main1.cpp (hunk 22 above)
Comment (st4rl3ss) [reply coreneuron/nrniv/main1.cpp]: I'll switch back to the old form then!

File: coreneuron/nrniv/corenrn_parameters.cpp (hunk 21 above)
Comment (pramodk) [reply coreneuron/nrniv/corenrn_parameters.cpp]: @ohm314 : `CLI::ParseError` gives access to name of the error and error message itself:

```
The following argument was not expected: -mpi
```

we can parse that error message. But, we can instead go through argv and check if old argument like `-mpi` is used.

Comment (ohm314) [reply coreneuron/nrniv/corenrn_parameters.cpp]: I dug into the CLI11 exception code and indeed there seems to be no nice way to simply access those bad arguments. Then I guess for now we can rely on CLI11 returning its error message.

Comment (ohm314) [reply coreneuron/nrniv/corenrn_parameters.cpp]: I would maybe add additionally add something like:
```
cerr << "Single-dash arguments such as -mpi are depricated, please check ./coreneuron --help for more information." << std::endl
```

**Comment by pramodk** [comment]
This looks fine. My only question is when to trigger the merge. As this changes all command line option, before merge we have to make sure following things are updated:

* https://github.com/nrnhines/ringtest
* `lib/modlib/CoreConfig.mod` in old neurodamus as well new neurodamus

File: README.md [hunk 27]
```diff
@@ -132,63 +132,102 @@ cmake .. -DCMAKE_CXX_FLAGS="-O3 -g" \
...
---multisend                 Use Multisend spike exchange instead of Allgather.
---read-config ARG           Read configuration file filename.
---restore ARG               Restore simulation from provided checkpoint
-                            directory.
---show                      Print args.
---skip-mpi-finalize         Do not call mpi finalize.
---spkcompress ARG           Spike compression. Up to ARG are exchanged during
-                            MPI_Allgather. (0)
---write-config ARG          Write configuration file filename.
+/path/to/install/directory/coreneuron_exec --help-all
```
Comment (alkino) [inline README.md]: I plan to remove the print of the help in README?

Any thoughts?

Comment (st4rl3ss) [reply README.md]: Nothing against it here, actually it makes sense!

**Comment by alkino** [comment]
How should we handle `blueconfigs/scx-v5-plasticity/output/sim.conf` for example, producing old parameters scheme?

**Comment by alkino** [comment]
I reduce impact of this patch.

**Comment by alkino** [comment]
Please retest

**Comment by alkino** [comment]
Please retest

File: tests/jenkins/run_neuron.sh [hunk 28]
```diff
@@ -13,12 +13,12 @@ cd $WORKSPACE/${TEST_DIR}
 if [ "${TEST_DIR}" = "testcorenrn" ]; then
     mkdir test${TEST}dat
     mkdir ${TEST}
-    mpirun -n ${MPI_RANKS} ./x86_64/special --mpi -c sim_time=100 test${TEST}.hoc
+    mpirun -n ${MPI_RANKS} ./x86_64/special -mpi -c sim_time=100 test${TEST}.hoc
```
Comment (ferdonline) [inline tests/jenkins/run_neuron.sh]: I know this was already there, but `mpiexec` is guaranteed to exist in mpi implementations, not mpirun. Besides there were these CPU affinity problem with mpirun.

Comment (alkino) [reply tests/jenkins/run_neuron.sh]: #250

File: coreneuron/apps/corenrn_parameters.cpp [hunk 29]
```diff
@@ -0,0 +1,178 @@
...
+    CLI::retire_option(app, "--show");
+};
+
+void corenrn_parameters::parse (int argc, char** argv) {
+
+    try {
+        app.parse(argc, argv);
+    } catch (const CLI::ExtrasError &e) {
+
+        std::cerr << "Single-dash arguments such as -mpi are deprecated, please check ./coreneuron_exec --help for more information. \n" << std::endl;
```
Comment (pramodk) [inline coreneuron/apps/corenrn_parameters.cpp]:
```suggestion
        std::cerr << "Single-dash arguments such as -mpi are removed, please check ./coreneuron_exec --help for more information. \n" << std::endl;
```
it's not just deprecated but removed, isn't it?

File: tests/jenkins/run_corenrn.sh [hunk 30]
```diff
@@ -18,13 +18,12 @@ else
 fi
 
 if [ "${TEST}" = "patstim" ]; then
-    mpirun -n ${MPI_RANKS} ./${CORENRN_TYPE}/special-core --pattern patstim.spk -mpi -d test${TEST}dat -e 100 -o ${TEST}
+    mpirun -n ${MPI_RANKS} ./${CORENRN_TYPE}/special-core --show --mpi -e 100 --pattern patstim.spk -d test${TEST}dat -o ${TEST}
```
Comment (pramodk) [inline tests/jenkins/run_corenrn.sh]: Isn't --show is default?

**Comment by alkino** [comment]
Do it à la boost.

0.16.0 => 001600
1.2.3 => 010203

Easy to compare

**Comment by ferdonline** [comment]
Exactly! But I left patch out, I dont expect that level of granularity to be required. But what do you think?

File: coreneuron/engine.h.in [hunk 31]
```diff
@@ -2,19 +2,19 @@
...
-#define CORENEURON_VERSION @coreneuron_VERSION_MAJOR@.@coreneuron_VERSION_MINOR@
+#define CORENEURON_VERSION @CORENEURON_VERSION_COMBINED@
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 /// All-in-one initialization of mechanisms and solver
-int solve_core(int argc, char** argv);
+extern int solve_core(int argc, char** argv);
```
Comment (ferdonline) [inline coreneuron/engine.h.in]: extern is apparently required in C99 mode 😫

**Comment by alkino** [comment]
So let's go ;)
